src/Aspire.Dashboard/Model/ResourceOutgoingPeerResolver.cs

```diff
@@ -0,0 +1,133 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Concurrent;
+
+namespace Aspire.Dashboard.Model;
+
+public sealed class ResourceOutgoingPeerResolver : IOutgoingPeerResolver, IAsyncDisposable
+{
+    private readonly IDashboardViewModelService _dashboardViewModelService;
+    private readonly ConcurrentDictionary<string, ResourceViewModel> _resourceNameMapping = new();
+    private readonly CancellationTokenSource _watchContainersTokenSource = new();
+    private readonly Task _watchTask;
+    private readonly List<Subscription> _subscriptions;
+    private readonly object _lock = new object();
+
+    public ResourceOutgoingPeerResolver(IDashboardViewModelService dashboardViewModelService)
+    {
+        _dashboardViewModelService = dashboardViewModelService;
+        _subscriptions = new List<Subscription>();
+
+        var viewModelMonitor = _dashboardViewModelService.GetResources();
+        var initialList = viewModelMonitor.Snapshot;
+        var watch = viewModelMonitor.Watch;
+
+        foreach (var result in initialList)
+        {
+            _resourceNameMapping[result.Name] = result;
+        }
+
+        _watchTask = Task.Run(async () =>
+        {
+            await foreach (var resourceChanged in watch.WithCancellation(_watchContainersTokenSource.Token))
+            {
+                await OnResourceListChanged(resourceChanged.ObjectChangeType, resourceChanged.Resource).ConfigureAwait(false);
+            }
+        });
+    }
+
+    private async Task OnResourceListChanged(ObjectChangeType changeType, ResourceViewModel resourceViewModel)
+    {
+        if (changeType == ObjectChangeType.Added)
+        {
+            _resourceNameMapping[resourceViewModel.Name] = resourceViewModel;
+        }
+        else if (changeType == ObjectChangeType.Modified)
+        {
+            _resourceNameMapping[resourceViewModel.Name] = resourceViewModel;
+        }
+        else if (changeType == ObjectChangeType.Deleted)
+        {
+            _resourceNameMapping.TryRemove(resourceViewModel.Name, out _);
```

Comment:
Hey @JamesNK are we sure this is safe? This modification of the collection could coincide with the enumeration below in the public `ResolvePeerName` method, which would lead to a collection modification exception at runtime. Looks like there needs to be some more protection here.

Reply:
It's not safe. Note that I copied this from the console logs page and the same thing happens on the resources page.

I'd like to see getting resources simplified. The code around it is very complicated.

Reply:
I'll simplify some of it while working on the data transport.

Filed https://github.com/dotnet/aspire/issues/1239 to track the above exception.